phoenix-core-client/src/main/java/org/apache/phoenix/util/ValidateLastDDLTimestampUtil.java

```diff
@@ -0,0 +1,231 @@
+        } catch (Exception e) {
+            SQLException parsedException = ClientUtil.parseServerException(e);
+            if (parsedException instanceof StaleMetadataCacheException) {
+                throw parsedException;
+            }
+            //retry once for any exceptions other than StaleMetadataCacheException
+            LOGGER.error("Error in validating DDL timestamp for {}", infoString, parsedException);
+            if (doRetry) {
+                // update the list of live region servers
+                conn.getQueryServices().refreshLiveRegionServers();
+                validateLastDDLTimestamp(conn, tableRefs, false);
+                return;
+            }
+            throw parsedException;
+        }
+    }
+
+    /**
+     * Build a request for the validateLastDDLTimestamp RPC for the given tables.
+     * 1. For a view, we need to add all its ancestors to the request
+     * in case something changed in the hierarchy.
+     * 2. For an index, we need to add its parent table to the request
+     * in case the index was dropped.
+     * 3. Add all indexes of a table/view in case index state was changed.
+     *
+     * @param tableRefs
+     * @return ValidateLastDDLTimestampRequest for the table in tableRef
+     */
+    private static RegionServerEndpointProtos.ValidateLastDDLTimestampRequest
+        getValidateDDLTimestampRequest(List<TableRef> tableRefs) {
+
+        RegionServerEndpointProtos.ValidateLastDDLTimestampRequest.Builder requestBuilder
+                = RegionServerEndpointProtos.ValidateLastDDLTimestampRequest.newBuilder();
+        RegionServerEndpointProtos.LastDDLTimestampRequest.Builder innerBuilder;
+
+        for (TableRef tableRef : tableRefs) {
+
+            // validate all ancestors of this PTable if any
+            // index -> base table
+            // view -> parent view and its ancestors
+            // view index -> view and its ancestors
+            for (Map.Entry<PTableKey, Long> entry
+                    : tableRef.getTable().getAncestorLastDDLTimestampMap().entrySet()) {
+                innerBuilder = RegionServerEndpointProtos.LastDDLTimestampRequest.newBuilder();
+                PTableKey ancestorKey = entry.getKey();
+                setLastDDLTimestampRequestParameters(innerBuilder, ancestorKey, entry.getValue());
+                requestBuilder.addLastDDLTimestampRequests(innerBuilder);
+            }
+
+            // add the current table to the request
+            PTable ptable = tableRef.getTable();
+            innerBuilder = RegionServerEndpointProtos.LastDDLTimestampRequest.newBuilder();
+            setLastDDLTimestampRequestParameters(innerBuilder, ptable.getKey(),
+                    ptable.getLastDDLTimestamp());
+            requestBuilder.addLastDDLTimestampRequests(innerBuilder);
+
+            // add all indexes of the current table
+            for (PTable idxPTable : tableRef.getTable().getIndexes()) {
```

Comment:
When validating timestamps we validate the timestamps of all the ancestors and the current table as well as indexes. When the server returns StaleMetadataCacheException are we also re-populating the client cache for all the objects that we are validating ? Are we also re-populating the indexes of the current table ?

Reply:
Yes. Currently, we are re-populating everything since we use `MetaDataClient.updateCache` workflow - it resolves ancestors and goes through the process of adding inherited columns, indexes from them, it also re-populates the indexes into the client cache.